File: 36-ValidSudoku/36-ValidSudoku.py
# Last updated: 8/1/2025, 11:40:47 PM
from collections import defaultdict
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        cols = defaultdict(set)
        row = defaultdict(set)
        box = defaultdict(set)
        for r in range(9):
            for c in range(9):
                val = board[r][c]

                if val == '.':
                    continue

            
                if (val in row[r] or 
                    val in cols[c] or 
                    val in box[(r // 3, c // 3)]):
                    return False

            
            
                cols[c].add(board[r][c])
                row[r].add(board[r][c])
                box[r//3,c//3].add(board[r][c])

        return True
            
        

    

File: 36-ValidSudoku/36-ValidSudoku.cpp
// Last updated: 8/30/2025, 11:52:16 AM
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        vector<set<char>> rowset(9);
        vector<set<char>> colset(9);
        map<pair<int, int>,set<char>> box;
        for (int r = 0; r < 9; r++) {
            for (int c = 0; c < 9; c++) {
                if (board[r][c] == '.') {
                    continue;
                }
                if ((rowset[r].count(board[r][c])) ||
                    (colset[c].count(board[r][c])) || (box[{r / 3, c / 3}].count(
                                                       board[r][c]))) {
                                                    return false;
                                                        
                }
                rowset[r].insert(board[r][c]);
                colset[c].insert(board[r][c]);
                box[{r/3,c/3}].insert(board[r][c]);
            }
        }
        return true;
    }
};